common/render.go

```diff
@@ -153,63 +155,46 @@ func (r renderEntityList) Len() int {
 
 func (r renderEntityList) Less(i, j int) bool {
 	// Sort by shader-pointer if they have the same zIndex
-	if r[i].RenderComponent.zIndex == r[j].RenderComponent.zIndex {
-		// // TODO: optimize this for performance
-		p1 := fmt.Sprintf("%p", r[i].RenderComponent.shader)
-		p2 := fmt.Sprintf("%p", r[j].RenderComponent.shader)
-		// Sort by texture if shader is the same
-		// fmt.Println(p1, p2)
-		if p1 == p2 {
-			switch r[i].RenderComponent.Drawable.(type) {
-			// Tiles can either be as a spriteSheet or as separate image
-			// if we sort them by texture and they're saved as separate images,
-			// sorting by texture messes up rendering.
-			case *Tile:
-				mag1 := r[i].RenderComponent.magFilter
-				mag2 := r[j].RenderComponent.magFilter
-				// Sort by minFilter if they're the same magFilter
-				if mag1 == mag2 {
-					min1 := r[i].RenderComponent.minFilter
-					min2 := r[j].RenderComponent.minFilter
-					// Sort by position if they're the same minFilter
-					if min1 == min2 {
-						if r[i].Position.Y == r[j].Position.Y {
-							return r[i].Position.X < r[j].Position.X
-						}
-						return r[i].Position.Y < r[j].Position.Y
-					}
-					return min1 < min2
-				}
-				return mag1 < mag2
-			default:
-				t1 := fmt.Sprintf("%p", r[i].RenderComponent.Drawable.Texture())
-				t2 := fmt.Sprintf("%p", r[j].RenderComponent.Drawable.Texture())
-				// Sort by magFilter if they're the same texture
-				if t1 == t2 {
-					mag1 := r[i].RenderComponent.magFilter
-					mag2 := r[j].RenderComponent.magFilter
-					// Sort by minFilter if they're the same magFilter
-					if mag1 == mag2 {
-						min1 := r[i].RenderComponent.minFilter
-						min2 := r[j].RenderComponent.minFilter
-						// Sort by position if they're the same minFilter
-						if min1 == min2 {
-							if r[i].Position.Y == r[j].Position.Y {
-								return r[i].Position.X < r[j].Position.X
-							}
-							return r[i].Position.Y < r[j].Position.Y
-						}
-						return min1 < min2
-					}
-					return mag1 < mag2
-				}
-				return t1 < t2
-			}
-		}
+	if r[i].RenderComponent.zIndex != r[j].RenderComponent.zIndex {
+		return r[i].RenderComponent.zIndex < r[j].RenderComponent.zIndex
+	}
+
+	// TODO: optimize this for performance (is it enough? maybe "cache" shader uintptr (i.e: r.shaderPtr)?)
+	p1, p2 := (*[2]uintptr)(unsafe.Pointer(&r[i].RenderComponent.shader))[1], (*[2]uintptr)(unsafe.Pointer(&r[j].RenderComponent.shader))[1]
```

Comment:
Actually, there's already a shader comparison function `compareShaders` that does that during the draw step. It's optimized for speed and also has a version that works for gopherjs and wasm. We should just use that here XP

Reply:
The _problem_ is that `compareShaders` just return a bool, but not the `pointer` (we need to compare `p1 < p2`). Maybe a new `compareShadersLess(a, b Shader) bool` 🤣. In that case if it's true it's not equal and just return.

```
p := comapreSharersLess(r[i].RenderComponent.shader, r[j].RenderComponent.shader)
if p {
    return p
}
```
```
func compareShadersLess(a, b Shader) bool {
	aIface := *(*iface)(unsafe.Pointer(&a))
	bIface := *(*iface)(unsafe.Pointer(&b))
	return uintptr(aIface.Data) < uintptr(bIface.Data)
}
```
😅

Reply:
Alright, but the way it's done now breaks gopherjs and wasm builds. To fix this I propose we add to render_notjs.go a `lessShaders(a, b Shader) bool` that will return wether shader a is less than shader b using the unsafe package, and then a similar function in render_js.go that does the same thing without using unsafe. Then call that at the return for `p1 < p2`.

We can do this in one of two ways:
1) I can merge this as-is and then fix it in a later commit myself

2) You can fix it here, if you have time and want to do it

either way works for me ^_^

Reply:
I think I fix that. But has another issue. I'm testing the `entityscroller.go`, it doesn't render because of `StartZIndex`. I don't know why, but `render.zIndex = render.StartZIndex` breaks the game. The only way to fix seems to be:

```
func (r *RenderComponent) SetZIndex(index float32) {
	r.StartZIndex = index
	r.zIndex = index
	engo.Mailbox.Dispatch(&renderChangeMessage{})
}
```

However, how the same component is added again? 😕

Reply:
Ok. I found the issue.